ledger/src/shred/merkle.rs

```diff
@@ -0,0 +1,523 @@
+        }
+        if !self.verify_merkle_proof()? {
+            return Err(Error::InvalidMerkleProof);
+        }
+        shred_data::sanitize(self)
+    }
+
+    fn signed_message(&self) -> &[u8] {
+        self.merkle_branch.root.as_ref()
+    }
+}
+
+impl Shred for ShredCode {
+    impl_shred_common!();
+    const SIZE_OF_PAYLOAD: usize = shred_code::ShredCode::SIZE_OF_PAYLOAD;
+
+    fn from_payload(mut payload: Vec<u8>) -> Result<Self, Error> {
+        let mut cursor = Cursor::new(&payload[..]);
+        let common_header: ShredCommonHeader = deserialize_from_with_limit(&mut cursor)?;
+        let proof_size = match common_header.shred_variant {
+            ShredVariant::MerkleCode(proof_size) => proof_size,
+            _ => return Err(Error::InvalidShredVariant),
+        };
+        let coding_header = deserialize_from_with_limit(&mut cursor)?;
+        // Skip erasure code shard.
+        let shard_size = Self::size_of_erasure_encoded_slice(proof_size)?;
+        let shard_size = i64::try_from(shard_size).unwrap();
+        cursor.seek(SeekFrom::Current(shard_size))?;
+        // Deserialize merkle branch.
+        let root = deserialize_from_with_limit(&mut cursor)?;
+        let proof = repeat_with(|| deserialize_from_with_limit(&mut cursor))
+            .take(usize::from(proof_size))
+            .collect::<Result<_, _>>()?;
+        let merkle_branch = MerkleBranch { root, proof };
+        // see: https://github.com/solana-labs/solana/pull/10109
+        payload.truncate(Self::SIZE_OF_PAYLOAD);
+        let shred = Self {
+            common_header,
+            coding_header,
+            merkle_branch,
+            payload,
+        };
+        shred.sanitize().map(|_| shred)
+    }
+
+    fn erasure_shard_index(&self) -> Result<usize, Error> {
+        shred_code::erasure_shard_index(self).ok_or_else(|| {
+            let headers = Box::new((self.common_header, self.coding_header));
+            Error::InvalidErasureShardIndex(headers)
+        })
+    }
+
+    fn erasure_shard(self) -> Result<Vec<u8>, Error> {
+        if self.payload.len() != Self::SIZE_OF_PAYLOAD {
+            return Err(Error::InvalidPayloadSize(self.payload.len()));
+        }
+        let proof_size = self.proof_size()?;
+        let shard_size = Self::size_of_erasure_encoded_slice(proof_size)?;
+        let mut shard = self.payload;
+        shard.drain(..SIZE_OF_CODING_SHRED_HEADERS);
+        shard.truncate(shard_size);
+        Ok(shard)
+    }
+
+    fn erasure_shard_as_slice(&self) -> Result<&[u8], Error> {
+        if self.payload.len() != Self::SIZE_OF_PAYLOAD {
+            return Err(Error::InvalidPayloadSize(self.payload.len()));
+        }
+        let proof_size = self.proof_size()?;
+        let shard_size = Self::size_of_erasure_encoded_slice(proof_size)?;
+        self.payload
+            .get(SIZE_OF_CODING_SHRED_HEADERS..SIZE_OF_CODING_SHRED_HEADERS + shard_size)
+            .ok_or(Error::InvalidPayloadSize(self.payload.len()))
```

Comment:
checked by `size_of_erasure_encoded_slice`?

Reply:
for now using the safe api until the code is in steady state